File: Medium/python/1325. Delete Leaves With a Given Value.py
"""
Given a binary tree root and an integer target, delete all the leaf nodes with value target.

Note that once you delete a leaf node with value target, if its parent node becomes a leaf node and has the value target, it should also be deleted (you need to continue doing that until you cannot).
"""

from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def removeLeafNodes(self, root: Optional[TreeNode], target: int) -> Optional[TreeNode]:
        if not root:
            return None
        if not root.left and not root.right:
            return None if root.val == target else root

        left = self.removeLeafNodes(root.left, target=target)
        right = self.removeLeafNodes(root.right, target=target)

        if left == None and right == None:
            root.left = None
            root.right = None
            return None if root.val == target else root
        if left == None:
            root.left = None
        elif right == None:
            root.right = None

        return root

x = TreeNode(8)
print(Solution().removeLeafNodes(x, 8))

File: Medium/Javascript/1325. Delete Leaves With a Given Value.js
/**Given a binary tree root and an integer target, delete all the leaf nodes with value target.

Note that once you delete a leaf node with value target, if its parent node becomes a leaf node and has the value target, it should also be deleted (you need to continue doing that until you cannot).
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} target
 * @return {TreeNode}
 */
var removeLeafNodes = function (root, target) {
  if (!root) {
    return null;
  }
  if (!root.left && !root.right) {
    return target === root.val ? null : root;
  }

  let left = removeLeafNodes(root.left, target);
  let right = removeLeafNodes(root.right, target);

  if (!left && !right) {
    root.left = null;
    root.right = null;
    return target === root.val ? null : root;
  }
  if (!left) {
    root.left = null;
  } else if (!right) {
    root.right = null;
  }

  return root;
};
